design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+# Branch Checks
+
+## User Story
+
+In order to achieve CI/CD for data, data engineering teams would like to validate the data itself before allowing to merge into a production branch.
+These validations might take longer than a typical HTTP request.
+
+## Goals
+
+Currently, lakeFS hooks allow running `pre-*` type hooks as blocking operations: They execute in a manner that is tied to an HTTP request.
+This puts a pretty short time window for hooks to complete in.
+
+Distributed processing frameworks such as Apache Spark typically have a long startup time and might take minutes (or even hours) to complete.
+
+The main goal of this proposal is to allow such validations to occur with a very long timeout (configurable, but could be days).
+
+## Non-Goals
+
+Out of scope for this proposal is any collaboration features that are typically associated with PRs (that provide similar functionality to the one being proposed).
+
+## High Level design
+
+This proposal adds a `Check` primitive to lakeFS branches. a Check is similar to a hook: it calls out to some external component with information about a modification or event that occurred in the system.
+
+The user flow for checks is the following:
+
+1. configure checks
+1. ensure all *mandatory* checks have passed for the given branch's HEAD
+1. merge said branch into a protected one.
+
+Merging into any protected branch requires all mandatory checks to have completed successfully for the current HEAD:
+
+```mermaid
+sequenceDiagram
+    Actor user1
+    user1->>lakeFS: commit1(branch_a)
+    user1->>lakeFS: commit2(branch_a)
+    user1->>lakeFS: commitN(branch_a)
+    user1->>lakeFS: check(branch_a) // will be dereferenced
+    loop For every configured check
+      lakeFS->>+External Executor: execute(event details, callback URL)
+      External Executor->>-lakeFS: callback(status)
+      lakeFS->>lakeFS: update check status (SUCCESS)
+    end
+    user1->>lakeFS: merge(branch_a, branch_b_protected)
+    lakeFS-->>user1: success!
+```
+
+## Implementation
+
+#### Check configuration
+
+Checks are configured under `_lakefs_actions/` within a lakeFS repository's root. Any YAML file within the directory can contain checks, similarly to hooks.
```

Comment:
Suggest separating checks and hooks files either by a different path or a subdir. If it was up to me, I would allow only a single checks file with a specific path.
Putting the checks and hooks together creates an overhead of reading the entire actions path to filter only for the checks. You can have dozen of action files which may contain only a single check in one of them and you will need to read all the files from storage and parse them.

Reply:
Allowing users to decide how to manage these files is a property I'd like to retain. I can share a use case from existing users where they provision repositories automatically and have a process that writes a couple of post-commit hooks into `_lakefs_actions/`. 

Because each file is independent, automating such processes is easy - no need to read or parse the existing checks. AFAIK lakeFS doesn't care (or at least, shouldn't care) if objects under `_lakefs_actions/` include a `/` delimiter - so users can decide - and we can even recommend, that they store checks under `_lakefs_actions/checks/` and hooks under `_lakefs_actions/hooks/<event-type>/` to make it easier for humans to consume.